File: runtimes/js/src/capabilities.ts
type CapabilityMode = "read" | "write";

interface ParsedCapability {
	mode: string;
	scope: string;
}

function splitMode(cap: string): ParsedCapability {
	const i = cap.indexOf(":");
	if (i < 0) {
		return { mode: "write", scope: cap };
	}
	return { mode: cap.slice(0, i), scope: cap.slice(i + 1) };
}

function modeSatisfies(grantMode: string, requireMode: string): boolean {
	return grantMode === "write" || grantMode === requireMode;
}

function scopeSubsumes(grantScope: string, requireScope: string): boolean {
	return grantScope === requireScope || requireScope.startsWith(grantScope + ".");
}

export function satisfies(grant: string, require: string): boolean {
	const g = splitMode(grant);
	const r = splitMode(require);
	return modeSatisfies(g.mode, r.mode) && scopeSubsumes(g.scope, r.scope);
}

export function grantedSatisfies(granted: Set<string> | Iterable<string>, require: string): boolean {
	for (const g of granted) {
		if (satisfies(g, require)) return true;
	}
	return false;
}

export type { CapabilityMode };

File: runtimes/csharp/src/Xript.Runtime/Capabilities.cs
namespace Xript.Runtime;

internal static class Capabilities
{
    internal static bool Satisfies(string grant, string require)
    {
        var (grantMode, grantScope) = SplitMode(grant);
        var (requireMode, requireScope) = SplitMode(require);

        var modeOk = grantMode == "write" || grantMode == requireMode;
        var scopeOk = grantScope == requireScope
            || requireScope.StartsWith(grantScope + ".", StringComparison.Ordinal);

        return modeOk && scopeOk;
    }

    internal static bool GrantedSatisfies(IEnumerable<string> granted, string require)
    {
        foreach (var grant in granted)
            if (Satisfies(grant, require))
                return true;
        return false;
    }

    private static (string Mode, string Scope) SplitMode(string capability)
    {
        var i = capability.IndexOf(':');
        if (i < 0)
            return ("write", capability);

        var prefix = capability[..i];
        return (prefix, capability[(i + 1)..]);
    }
}
